### PR TITLE
Issue 88

### PR DESCRIPTION
**Pull request by azubieta**
fixing clash between libusb and qt events loop in linux 
:smile:

### CODE REVIEW

File: src/AppGui.cpp [hunk 1]
```diff
@@ -89,8 +89,12 @@ bool AppGui::initialize()
     QMenu *systrayMenu = new QMenu();
 
     daemonAction = new DaemonMenuAction(systrayMenu);
-    systrayMenu->addAction(daemonAction);
-    systrayMenu->addSeparator();
+    // FIXME: Shall we add a fallback for linux and MacOs?
+    #ifdef Q_OS_WIN
```
Comment (raoulh) [inline src/AppGui.cpp]: No. This works on macos, it's only on linux that does not support widgets in QMenu

File: src/UsbMonitor_linux.cpp [hunk 2]
```diff
@@ -105,62 +146,113 @@ void UsbMonitor_linux::start()
...
+    qDeleteAll(m_fdMonitors.values());
+    m_fdMonitors.clear();
 }
 
-UsbMonitor_linux::~UsbMonitor_linux()
+void LibUSBWorker::createSocketMonitor(int fd)
 {
-    stop();
-    libusb_exit(usb_ctx);
+    if (!m_fdMonitors.contains(fd)) {
```
Comment (raoulh) [inline src/UsbMonitor_linux.cpp]: } on newline please

File: src/UsbMonitor_linux.cpp [hunk 3]
```diff
@@ -105,62 +146,113 @@ void UsbMonitor_linux::start()
...
+        QSocketNotifier* sn = new QSocketNotifier(fd, QSocketNotifier::Read);
+        connect(sn, &QSocketNotifier::activated, [=]() {
+//            qDebug() << "Event received";
+
+            if (libusb_try_lock_events(usb_ctx) == 0)
+            {
+                if (!libusb_event_handling_ok(usb_ctx)) {
```
Comment (raoulh) [inline src/UsbMonitor_linux.cpp]: } on newline please

File: src/UsbMonitor_linux.cpp [hunk 4]
```diff
@@ -105,62 +146,113 @@ void UsbMonitor_linux::start()
...
 }
 
-UsbMonitor_linux::~UsbMonitor_linux()
+void LibUSBWorker::createSocketMonitor(int fd)
 {
-    stop();
-    libusb_exit(usb_ctx);
+    if (!m_fdMonitors.contains(fd)) {
+        QSocketNotifier* sn = new QSocketNotifier(fd, QSocketNotifier::Read);
+        connect(sn, &QSocketNotifier::activated, [=]() {
```
Comment (raoulh) [inline src/UsbMonitor_linux.cpp]: } on newline please

File: src/UsbMonitor_linux.cpp [hunk 5]
```diff
@@ -105,62 +146,113 @@ void UsbMonitor_linux::start()
...
+                }
+            }
+        });
+        m_fdMonitors[fd] = sn;
+    }
+}
+
+void LibUSBWorker::releaseSocketMonitor(int fd)
+{
+    if (m_fdMonitors.contains(fd)) {
```
Comment (raoulh) [inline src/UsbMonitor_linux.cpp]: } on newline please

File: src/UsbMonitor_linux.cpp [hunk 6]
```diff
@@ -85,16 +92,50 @@ UsbMonitor_linux::UsbMonitor_linux()
...
+{
+    emit stopWorker();
+    relaseCallbacks();
+}
+
+void UsbMonitor_linux::handleEvents()
+{
+//    qDebug() << "Handling events in " << QThread::currentThread();
+    int completed = 0;
+    while (!completed){
```
Comment (raoulh) [inline src/UsbMonitor_linux.cpp]: } on newline please

File: src/AppGui.cpp (hunk 1 above)
Comment (raoulh) [reply src/AppGui.cpp]: Also on linux, normal qmenu entries could be added for starting/stopping the daemon

**Comment by limpkin** [comment]
Didn't work for me on Ubuntu

On Sep 12, 2017 8:55 AM, "raoulh" <notifications@github.com> wrote:

*@raoulh* commented on this pull request.
------------------------------

In src/AppGui.cpp
<https://github.com/raoulh/moolticute/pull/94#discussion_r138265129>:

> @@ -89,8 +89,12 @@ bool AppGui::initialize()
     QMenu *systrayMenu = new QMenu();

     daemonAction = new DaemonMenuAction(systrayMenu);
-    systrayMenu->addAction(daemonAction);
-    systrayMenu->addSeparator();
+    // FIXME: Shall we add a fallback for linux and MacOs?
+    #ifdef Q_OS_WIN

Also on linux, normal qmenu entries could be added for starting/stopping
the daemon

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/raoulh/moolticute/pull/94#discussion_r138265129>, or mute
the thread

**Comment by azubieta** [comment]
Will make a new PR, with the code rebased on current master HEAD.

**Pull request closed without merging**